beluga/test/beluga/policies/test_policy.cpp

```diff
@@ -0,0 +1,114 @@
+
+#include <beluga/policies/policy.hpp>
+
+namespace {
+
+using testing::Return;
+
+struct MockPredicate {
+  MOCK_METHOD(bool, eval, ());
+  bool operator()() { return eval(); }
+  auto policy() { return beluga::make_policy_closure(std::ref(*this)); }
+};
+
+TEST(Policy, AndTrue) {
+  MockPredicate condition1;
+  MockPredicate condition2;
+  auto combined = condition1.policy() && condition2.policy();
+  EXPECT_CALL(condition1, eval()).WillOnce(Return(true));
+  EXPECT_CALL(condition2, eval()).WillOnce(Return(true));
+  ASSERT_EQ(combined(), true);
+}
+
+TEST(Policy, AndFalseShortCircuit) {
+  MockPredicate condition1;
+  MockPredicate condition2;
+  auto combined = condition1.policy() && condition2.policy();
+  EXPECT_CALL(condition1, eval()).WillOnce(Return(false));
+  EXPECT_CALL(condition2, eval()).Times(0);
+  ASSERT_EQ(combined(), false);
+}
+
+TEST(Policy, AndFalse) {
+  MockPredicate condition1;
+  MockPredicate condition2;
+  auto combined = condition1.policy() & condition2.policy();
+  EXPECT_CALL(condition1, eval()).WillOnce(Return(false));
+  EXPECT_CALL(condition2, eval()).WillOnce(Return(true));
+  ASSERT_EQ(combined(), false);
+}
+
+TEST(Policy, OrFalse) {
+  MockPredicate condition1;
+  MockPredicate condition2;
+  auto combined = condition1.policy() || condition2.policy();
+  EXPECT_CALL(condition1, eval()).WillOnce(Return(false));
+  EXPECT_CALL(condition2, eval()).WillOnce(Return(false));
+  ASSERT_EQ(combined(), false);
+}
+
+TEST(Policy, OrTrueShortCircuit) {
+  MockPredicate condition1;
+  MockPredicate condition2;
+  auto combined = condition1.policy() || condition2.policy();
+  EXPECT_CALL(condition1, eval()).WillOnce(Return(true));
+  EXPECT_CALL(condition2, eval()).Times(0);
+  ASSERT_EQ(combined(), true);
+}
+
+TEST(Policy, OrTrue) {
+  MockPredicate condition1;
+  MockPredicate condition2;
+  auto combined = condition1.policy() | condition2.policy();
+  EXPECT_CALL(condition1, eval()).WillOnce(Return(true));
+  EXPECT_CALL(condition2, eval()).WillOnce(Return(false));
+  ASSERT_EQ(combined(), true);
+}
+
+TEST(Policy, Not) {
+  MockPredicate condition;
+  auto negated = !condition.policy();
+  EXPECT_CALL(condition, eval()).WillOnce(Return(false));
+  ASSERT_EQ(negated(), true);
+  EXPECT_CALL(condition, eval()).WillOnce(Return(true));
+  ASSERT_EQ(negated(), false);
+}
+
+TEST(Policy, SameArgumentCombinations) {
+  auto is_even = beluga::make_policy_closure([](int n) { return n % 2 == 0; });
+  auto never = beluga::make_policy_closure([](int) { return false; });
+  auto policy1 = is_even || never;
+  ASSERT_EQ(policy1(2), true);
+  ASSERT_EQ(policy1(3), false);
+  auto policy2 = never || is_even;
+  ASSERT_EQ(policy2(2), true);
+  ASSERT_EQ(policy2(3), false);
+}
+
+TEST(Policy, DifferentArgumentCombinations) {
+  auto is_even = beluga::make_policy_closure([](int n) { return n % 2 == 0; });
+  auto always = beluga::make_policy_closure([]() { return true; });
+  auto policy1 = is_even && always;
```

Comment:
@nahueespinosa meta: it would be nice for `beluga::any_policy([](int n) {...})` to yield `beluga::any_policy<int>` so that we can do:

```suggestion
  auto is_even = beluga::any_policy([](int n) { return n % 2 == 0; });
  auto always = beluga::any_policy([]() { return true; });
  auto policy1 = is_even && always;
```

Reply:
Another thought. If this were the case, we could drop `any_` and have `beluga::policy` instead.

Reply:
std::function-like deduction guides for any_policy are really hard (maybe impossible) to implement, since our template class has variadic Args instead of specific signature type.

I took a look at this:
https://github.com/gcc-mirror/gcc/blob/a57439d61937925cec48df6166b2a805ae7054d5/libstdc%2B%2B-v3/include/bits/std_function.h#L722-L725

But the Signature is of a single type, while in our case we would need a variadic list.

Reply:
The llvm implementation is very similar:
https://github.com/llvm-mirror/libcxx/blob/78d6a7767ed57b50122a161b91f59f19c9bd0d19/include/functional#L2389

Some roadblocks I encountered when trying to do this:

- CTAD for alias templates is available in C++20 (workaround is defining a struct instead of an alias).
- Metafunction calls are not allowed to be directly used as the return type of a deduction guide.
- Metafunctions can't return first-class parameter packs.

The only alternative I can see is changing the any policy template parameters to be a single signature instead of a pack, using `any_policy<bool(Args...)>` instead of `any_policy<Args...>`. This is a bit redundant since all policies are expected to return bool.

Reply:
> Metafunction calls are not allowed to be directly used as the return type of a deduction guide.

Aha! That kills the over-engineered path I was ruminating. Nevermind, `make_policy` is just fine.